### PR TITLE
Handle `default` bootloader entry patterns correctly

### LINKED ISSUE #183: NixOS with systemd-boot broken in v0.4
This issue appears to have been introduced by 2e7e87b7.

NixOS, and probably other distros, place the full name of the default boot entry in the `default` option of `loader.conf`.

The above commit forcefully adds a further `.conf` to the name returned by `loader::default_entry_file()`, which results in `loader::load_default_entry()` attempting to open a bogus path such as `/loader/entries/foo.conf.conf`.

As far as I am aware, `/loader/loader.conf` is only a `systemd-boot` thing - i.e. it's not part of the [Boot Loader Spec], and is only defined in the [systemd manual].

According to that documentation, the `default` option should be handled as a glob pattern, to be matched against (presumably) the filenames in `/loader/entries/`.

I am happy to attempt a fix if others agree that this is a bug.

[Boot Loader Spec]: https://systemd.io/BOOT_LOADER_SPECIFICATION
[systemd manual]: https://www.freedesktop.org/software/systemd/man/loader.conf.html#

### PR DESCRIPTION
**Pull request by benmaddison**
Resolves #183

This update implements matching of boot entry names against the pattern in the
`default` option from `loader/loader.conf`, intended to be compatible with the
behaviour of `systemd-boot`.

## Outstanding work

- [ ] implement `[..]` style patterns - deferred. Tracked in #191
- [x] better error handling, fewer panics
- [x] discuss use of unstable `slice_take` feature

### CODE REVIEW

**Comment by rbradford** [comment]
@benmaddison Looks like a great start. I'm very happy with what you've done so far.

Some minor things:

1. Could you split the comment updates/reformatting into a separate commit
2. Try and follow the existing commit message style (with the starting "fat:" etc..)
3. Don't forget your "-s"

On the "slice_take" thing. I don't know ... do you think it's on a path for staying around or might be dropped...it seems quite a stale nightly feature with not much activity.

I would also prefer to fix the broken OS support right now than perhaps addressing missing feature gaps unless they are in active use (that could be in a follow-up!)

**Comment by benmaddison** [comment]
> Some minor things:
> 
>     1. Could you split the comment updates/reformatting into a separate commit
> 
>     2. Try and follow the existing commit message style (with the starting "fat:" etc..)
> 
>     3. Don't forget your "-s"

Yup - I plan to rebase and squash before asking for a final review

> On the "slice_take" thing. I don't know ... do you think it's on a path for staying around or might be dropped...it seems quite a stale nightly feature with not much activity.

I'm not sure that I know the answer to that. I'll take another look at it to see if I can re-implement that loop without loosing too much readability.

> I would also prefer to fix the broken OS support right now than perhaps addressing missing feature gaps unless they are in active use (that could be in a follow-up!)

Ack. Agreed.

**Comment by benmaddison** [comment]
Open issues for review:

- I have kept the use of `slice_take`: I'm not seeing a way of avoiding it without effectively re-implementing that feature here. I'm happy to commit to doing the re-factor in future if the feature is dropped or this project wants to move to stable before the feature is stabilized.
- `find_entry` will currently bail out with an error if a call to `compare_entry` fails (due to an unterminated string), rather than falling back to a valid entry that was found earlier. Making that happen is not hard, but will be verbose, and I'm not sure it's worth the bother since that can only really happen if the FAT FS is broken. Guidance welcome.

**Comment by rbradford** [comment]
I would love to test with this an integration test? Is there a cloud image I can use?

> Open issues for review:
> 
>     * I have kept the use of `slice_take`: I'm not seeing a way of avoiding it without effectively re-implementing that feature here. I'm happy to commit to doing the re-factor in future if the feature is dropped or this project wants to move to stable before the feature is stabilized.

:+1: 

> 
>     * `find_entry` will currently bail out with an error if a call to `compare_entry` fails (due to an unterminated string), rather than falling back to a valid entry that was found earlier. Making that happen is not hard, but will be verbose, and I'm not sure it's worth the bother since that can only really happen if the FAT FS is broken. Guidance welcome.

:+1: I don't think we need to worry about dealing with a corrupt FAT.

File: src/fat.rs [hunk 1]
```diff
@@ -12,11 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use core::convert::TryFrom;
+
```
Comment (rbradford) [inline src/fat.rs:16]: nit: Put all use statements together (old code might not but for new code this would be good.)

File: src/fat.rs [hunk 2]
```diff
@@ -863,11 +878,12 @@ impl<'a> Filesystem<'a> {
 
 #[cfg(test)]
 mod tests {
-    use super::Read;
-    use crate::part::tests::*;
     use std::convert::TryInto;
     use std::path::PathBuf;
 
+    use super::Read;
```
Comment (rbradford) [inline src/fat.rs:884]: nit: Ditto.

File: src/loader.rs [hunk 3]
```diff
@@ -160,9 +246,10 @@ pub fn load_default_entry(fs: &fat::Filesystem, info: &dyn boot::Info) -> Result
 
 #[cfg(test)]
 mod tests {
+    use core::convert::TryInto;
```
Comment (rbradford) [inline src/loader.rs:249]: nit: Ditto.

**Review by rbradford — COMMENTED** [review]
Really happy with this. Just some very minor nits.

**Comment by benmaddison** [comment]
I have `group_imports = StdExternalCrate` in my local fallback `rustfmt.toml` file, which was why that grouping was occurring.

To ensure that doesn't happen to future contributors (incl. me), I have dropped an empty `rustfmt.toml` into the repo root, so that the defaults get used.

**Comment by benmaddison** [comment]
> I would love to test with this an integration test? Is there a cloud image I can use?

The NixOS image that I was using when I found the issue is locally generated, and the downloadable ones ship with grub.

I've had a search around for other options. So far I've checked ubuntu, fedora and arch: they all use grub :-(

I think the best I can offer is to generate a minimal nixos one, and store it in the repo with LFS, but it will make cloning unpleasant.

**Comment by rbradford** [comment]
> I have `group_imports = StdExternalCrate` in my local fallback `rustfmt.toml` file, which was why that grouping was occurring.
> 
> To ensure that doesn't happen to future contributors (incl. me), I have dropped an empty `rustfmt.toml` into the repo root, so that the defaults get used.

Ah, cool.

**Review by rbradford — APPROVED** [review]

**Pull request merged**